egui/src/menu.rs

```diff
@@ -59,46 +72,464 @@ pub fn bar<R>(ui: &mut Ui, add_contents: impl FnOnce(&mut Ui) -> R) -> InnerResp
+        } else if (response.clicked() && !is_menu_open(root, id))
+            || (response.hovered() && root.is_some())
+        {
+            let pos = response.rect.left_bottom();
+            return MenuResponse::Create(pos, id);
+        } else if pointer.any_pressed() {
+            if let Some(pos) = pointer.interact_pos() {
+                if let Some(root) = root {
+                    let menu_state = root.menu_state.read().unwrap();
+                    let in_old_menu = menu_state.area_contains(pos);
+                    if (!in_old_menu && pointer.primary_down()) || root.id == id {
+                        return MenuResponse::Close;
+                    }
+                }
+            }
+        }
+        MenuResponse::Stay
+    }
+    fn context_interaction(
+        response: &Response,
+        root: &mut Option<MenuRoot>,
+        id: Id,
+    ) -> MenuResponse {
+        let response = response.interact(Sense::click());
+        let pointer = &response.ctx.input().pointer;
+        if pointer.any_pressed() {
+            if let Some(pos) = pointer.interact_pos() {
+                let mut destroy = false;
+                let mut in_old_menu = false;
+                if let Some(root) = root {
+                    let menu_state = root.menu_state.read().unwrap();
+                    in_old_menu = menu_state.area_contains(pos);
+                    destroy = root.id == response.id;
+                }
+                if !in_old_menu {
+                    let in_target = response.rect.contains(pos);
+                    if in_target && pointer.secondary_down() {
+                        return MenuResponse::Create(pos, id);
+                    } else if (in_target && pointer.primary_down()) || destroy {
+                        return MenuResponse::Close;
+                    }
+                }
+            }
+        }
+        MenuResponse::Stay
+    }
+    fn click_interaction_impl(
+        response: &Response,
+        root: &mut Option<MenuRoot>,
+        id: Id,
+        interaction: impl InteractionFn,
+    ) {
+        match interaction(response, root, id) {
+            MenuResponse::Create(pos, id) => {
+                *root = Some(MenuRoot::new(pos, id));
+            }
+            MenuResponse::Close => *root = None,
+            MenuResponse::Stay => {}
+        };
+    }
+    pub fn context_click_interaction(response: &Response, root: &mut Option<MenuRoot>, id: Id) {
+        Self::click_interaction_impl(response, root, id, Self::context_interaction)
+    }
+    pub fn stationary_click_interaction(response: &Response, root: &mut Option<MenuRoot>, id: Id) {
+        Self::click_interaction_impl(response, root, id, Self::stationary_interaction)
+    }
+}
+#[derive(Clone, PartialEq)]
+pub(crate) enum MenuResponse {
+    Close,
+    Stay,
+    Create(Pos2, Id),
+}
+impl MenuResponse {
+    pub fn is_close(&self) -> bool {
+        *self == Self::Close
+    }
+}
+pub(crate) enum EntryState {
```

Comment:
This was meant to disable hover visuals of menu items when a submenu was open. but because the menu ui may contain any kinds of widgets, this would need a different approach which fixes hover visuals for all widgets when a submenu is open.. The point of this is to prevent hovering highlights when the cursor crosses items while moving to an opened submenu (which may cross the parent menu without closing the submenu).